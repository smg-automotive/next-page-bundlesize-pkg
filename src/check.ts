/* eslint-disable @typescript-eslint/no-require-imports */
import path from 'path';
import fs from 'fs';
import { execSync } from 'child_process';

import {
  getPreviousConfig,
  writeNewConfigFile,
} from './externalConfigFileHandler';

const { parse } = require('yargs');
interface Args {
  maxSize: string;
  buildDir: string;
  delta: string;
  previousConfigFileName?: string;
}

interface Manifest {
  pages: {
    [pageName: string]: string[];
  };
  [key: string]: Record<string, string | string[]>;
}

export interface BundleSizeConfig {
  files: Array<{
    path: string;
    maxSize: string;
  }>;
}

const combineAppAndPageChunks = (manifest: Manifest, page: string) =>
  Array.from(
    new Set([...manifest.pages['/_app'], ...manifest.pages[page]]),
  ).filter((chunk) => chunk.match(/\.js$/));

const concatenatePageBundles = ({
  buildDir,
  manifest,
}: {
  buildDir: string;
  manifest: Manifest;
}): string[] =>
  Object.keys(manifest.pages).map((page) => {
    const firstLoadChunks = combineAppAndPageChunks(manifest, page).map(
      (chunk) => path.join(buildDir, chunk),
    );

    const outFile = path.join(
      buildDir,
      `.bundlesize${page.replace(/[/]/g, '_').replace(/[[\]]/g, '-')}`,
    );

    fs.writeFileSync(outFile, '');
    firstLoadChunks.forEach((chunk) => {
      const chunkContent = fs.readFileSync(chunk);
      fs.appendFileSync(outFile, chunkContent);
    });

    return outFile;
  });

const generateBundleSizeConfig = ({
  pageBundles,
  maxSize,
  previousConfiguration,
}: {
  pageBundles: string[];
  maxSize: string;
  previousConfiguration: BundleSizeConfig;
}): BundleSizeConfig => {
  const previousConfigurationMap = new Map(
    previousConfiguration.files.map((config) => [config.path, config.maxSize]),
  );
  return {
    files: pageBundles.map((pageBundleName) => ({
      path: pageBundleName,
      maxSize: previousConfigurationMap.get(pageBundleName) || maxSize,
    })),
  };
};

const extractArgs = (args: string[]) => {
  const parsedArgs = parse(args) as unknown as Args;
  const maxSize = parsedArgs.maxSize || '200 kB';
  const buildDir = parsedArgs.buildDir || '.next';
  const delta = parsedArgs.delta || '5 kB';

  return {
    maxSize,
    buildDir,
    delta,
    previousConfigFileName: parsedArgs.previousConfigFileName,
  };
};

export default function check(args: string[]) {
  try {
    const { maxSize, buildDir, delta, previousConfigFileName } =
      extractArgs(args);

    const manifestFile = path.join(buildDir, 'build-manifest.json');
    const manifest = JSON.parse(fs.readFileSync(manifestFile).toString());

    const pageBundles = concatenatePageBundles({ buildDir, manifest });
    const previousConfiguration = getPreviousConfig(
      buildDir,
      previousConfigFileName,
    );
    const config = generateBundleSizeConfig({
      pageBundles,
      maxSize,
      previousConfiguration,
    });
    const configFile = path.join(buildDir, 'next-page-bundlesize.config.json');
    fs.writeFileSync(configFile, JSON.stringify(config, null, 2));
    writeNewConfigFile(config, delta, maxSize, buildDir);

    execSync(`npx bundlesize --config=${configFile}`, { stdio: 'inherit' });
  } catch (err) {
    // eslint-disable-next-line no-console
    console.log(err);
    process.exit(1);
  }

  process.exit(0);
}
