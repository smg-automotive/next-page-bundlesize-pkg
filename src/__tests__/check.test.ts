import fs from 'fs';

import check from '../check';

const validRunConfig = [
  'jest',
  './node_modules/.bin/jest',
  '--maxSize',
  '1 kB',
  '--buildDir',
  './src/__tests__/.next',
];

describe('cli', () => {
  const mockExit = jest.spyOn(process, 'exit').mockImplementation();

  beforeEach(() => {
    jest.resetModules();
    jest.clearAllMocks();
  });

  it('produces config with correct files and sizes', () => {
    check(validRunConfig);

    const config = fs
      .readFileSync('./src/__tests__/.next/next-page-bundlesize.config.json')
      .toString();
    expect(config).toMatchSnapshot();
    expect(mockExit).toHaveBeenCalledWith(0);
  });

  it('produces concatenated bundle file', () => {
    check(validRunConfig);

    const page = fs
      .readFileSync('./src/__tests__/.next/.bundlesize_')
      .toString();
    expect(page).toMatchSnapshot();
    expect(mockExit).toHaveBeenCalledWith(0);
  });

  it('fails when bundles are larger than the limit', () => {
    check([
      'jest',
      './node_modules/.bin/jest',
      '--maxSize',
      '1 b',
      '--buildDir',
      './src/__tests__/.next',
    ]);
    expect(mockExit).toHaveBeenCalledWith(1);
  });

  it('uses the previous config if defined', () => {
    check([
      ...validRunConfig,
      '--previousConfigFileName',
      'master-config.json',
    ]);

    const updatedConfig = fs
      .readFileSync('./src/__tests__/.next/bundlesize.json')
      .toString();
    expect(updatedConfig).toMatchSnapshot();
  });

  it('adds a delta to the new config if smaller than maxSize', () => {
    check([
      ...validRunConfig,
      '--previousConfigFileName',
      'master-config.json',
      '--delta',
      '5 kB',
    ]);

    const updatedConfig = fs
      .readFileSync('./src/__tests__/.next/bundlesize.json')
      .toString();
    expect(updatedConfig).toMatchSnapshot();
  });
});
