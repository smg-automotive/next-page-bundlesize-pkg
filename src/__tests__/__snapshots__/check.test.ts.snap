// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cli adds a delta to the new config if smaller than maxSize 1`] = `
"{
  \\"files\\": [
    {
      \\"path\\": \\"src/__tests__/.next/.bundlesize_\\",
      \\"maxSize\\": \\"5.04KB\\"
    },
    {
      \\"path\\": \\"src/__tests__/.next/.bundlesize__app\\",
      \\"maxSize\\": \\"5.03KB\\"
    }
  ]
}"
`;

exports[`cli produces concatenated bundle file 1`] = `
"// _app.js
// index.js
"
`;

exports[`cli produces config with correct files and sizes 1`] = `
"{
  \\"files\\": [
    {
      \\"path\\": \\"src/__tests__/.next/.bundlesize_\\",
      \\"maxSize\\": \\"1 kB\\"
    },
    {
      \\"path\\": \\"src/__tests__/.next/.bundlesize__app\\",
      \\"maxSize\\": \\"1 kB\\"
    }
  ]
}"
`;

exports[`cli uses the previous config if defined 1`] = `
"{
  \\"files\\": [
    {
      \\"path\\": \\"src/__tests__/.next/.bundlesize_\\",
      \\"maxSize\\": \\"5.04KB\\"
    },
    {
      \\"path\\": \\"src/__tests__/.next/.bundlesize__app\\",
      \\"maxSize\\": \\"5.03KB\\"
    }
  ]
}"
`;
